### ISSUE TITLE
feature request: continuous-mode for large nozzles

### ISSUE THREAD

**Issue by gotthalmseder** [alexrj/Slic3r#4552]
I have lots of experience in printing large scale models with large nozzles (1mm, 1,2mm). 
Outdoor only PET is useabel for large objects because it sticks on the bed, does not shrink and is weather-resistant. 

But using large nozzles PET does a lot of stringing and artifacts occour because large nozzles loose filament in empty travel and air comes in.  Retraction does not help like it does with fine nozzles. 

So we need a slicing mode simular to the vase mode, but applicable to any shape. A mode, that reduces empty travel to a minimum and prints most models in a continuous string. Let us call it "continuous-mode". Continuous mode could allways be combined with 100% filling, because this mode ist most used in printing with large nozzles and there patern-filling is not a good idea. 

The pictures show typical artefacts that appear in large-nozzle-prints in layers where lots of empty travel occur. The slicer tells the printer to print the perimeters and  then afterwards the infill. The seperate filling after doing the perimeters is no good Idea for printing with large nozzles because filling distant areas leads to a lot of empty travel.  

The black and white graphic shows a solution. It shows what I mean with slicing in "continuous mode". 
This is a feature I miss so badly in any slicer that is on the market! 

![loesung](https://user-images.githubusercontent.com/43730156/46262140-74ca8700-c4fd-11e8-9081-0e9f8f189eca.jpg)
![p9182588](https://user-images.githubusercontent.com/43730156/46262141-74ca8700-c4fd-11e8-9d00-affa858ae3f4.jpg)
![problem](https://user-images.githubusercontent.com/43730156/46262142-75631d80-c4fd-11e8-884c-033438e8307a.jpg)
![zufluss](https://user-images.githubusercontent.com/43730156/46262144-75631d80-c4fd-11e8-9d8d-5f51812e2ecd.jpg)

**Comment by lordofhyphens** [comment]
Just as a FYI, what you are asking for has the possibility of taking forever to slice on a lot of geometries.

**Comment by gotthalmseder** [comment]
Yes, it could be something that lets slic3r overrule all other slicers. Because I never had any printing failures in objects with no empty travel. Whatever can be printed in vase mode is fine, even with the largest nozzle. 
A continuous mode would be a sort of vase-mode for any geometry. Empty travel should be just necessary when there are more than one areas, for example if you print two objects once. But in each object there should not be any empty travel. That would lead to the best results with large nozzles.

**Comment by gotthalmseder** [comment]
I spend 500€ for the programmer, who gives me this feature. But it has to be done in an intelligent way, so it does not need too much calculation-time. As I print a lot of geometric objects it would be fine if it copies pathways from the layer below, whereever the shape hasn't changed.

**Comment by lordofhyphens** [comment]
Vase mode is a hack in Slic3r; it slices normally and then connects the dots. 

So the basic requirements you are looking for are the following:

* For every island in a current slice, generate a toolpath that:
     * covers the entire island
     * is continuous
* Assume a single width. 
* For islands that contain bridges, operate normally.

**Comment by gotthalmseder** [comment]
What is an FYI ?

**Comment by lordofhyphens** [comment]
@gotthalmseder "For your information"

Islands in Slic3r refer to the area defined by some closed polygon. A cross-section of a pillar is one island.

**Comment by gotthalmseder** [comment]
But we need to distinguish between filled and unfilled islands. A two perimeter thick wall is an island, but it is unfilled. If the wall has thicker parts, they need to be filled before the perimeters are printed further on. Otherwise you get empty travel in the filling process. 
So my definition of a perimeter that gets near to itself is necessary. Near means it has to start a filling process before it closes. There has to be a gap left, to leave the filled area.

**Comment by supermerill** [comment]
idea:
Perimeters are loop.
We can open-close a loop almost a the same place.
We foloww the external perimeter. Each time we encouter a more in-depth perimeter, we connect to it and wait for it to return (recursively).
Should be easy enough to program.

But no infill.

**Comment by gotthalmseder** [comment]
Sounds good. Infinite Perimeters are a good solution for the most objects. But it seems we discuss here details without any programmer looking on our ideas.

**Comment by lordofhyphens** [comment]
@gotthalmseder supermerill is plenty capable.

**Comment by gotthalmseder** [comment]
Is he a programmer?

**Comment by gotthalmseder** [comment]
Thanks for this tipp. I am new in this forum. Most Forums do not have such member-sites. 
So you have allready done programms for slic3r. That's fine!

Are you interested in the feature I suggest?
As far as I could, I would support the devellopment. I am not a C++ programmer, but I have done programms in Java. So I am able to understand the logic of programming, discuss rules and I would be able to test the feature on different shapes to see if it solves the printing-problems I get by usinig large nozzles. My english is not the best, but it should be sufficient :-)

**Comment by gotthalmseder** [comment]
But: Not only fillings lead to empty travel. Slic3r also often skips Perimeters of closed islands and travels arround without sense: 
![unbenannt-2](https://user-images.githubusercontent.com/43730156/46489842-c4af9380-c806-11e8-835d-6d057dfc1326.jpg)

**Comment by supermerill** [comment]
It's because you select "thin walls"

**Comment by foreachthing** [comment]
I am a big fan of this "issue" right here!
@supermerill: IMHO the outer perimeter should not have an opening. If that gap is just a tiny bit under-filled your part will most likely break (or leak) right there. And, it leaves a visible line from top to bottom (assuming this happens at the same place for each layer).

**Comment by supermerill** [comment]
i know, just a little bug i corrected today.

**Comment by gotthalmseder** [comment]
Sorry, I have not been online this weekend. 
Supermerill, this is great! Is it allready testable?

**Comment by supermerill** [comment]
> filling-line (red)

These are gap-fill. You can enable/disable that, but it can't be merged in a continuous line with the perimeters. It's mathematically impossible (unless for some edge-cases).
To remove them, you can also change the wall thickness of your model and/or change the extrusion width.

slic3r print perimeters then infill then gap fill (if i remember well).

Here a test version (win32, i assume): https://mega.nz/#!9oRHTa6Y!7Dy_zJWrjegp1V2_bFRNgvNHMhZsYvXg-u2Wfv7TE8c
It's limited to 20 layers (because it's only to test if it's ok, not a real version) and thin walls aren't added.
There are a checkbox in the infill category to enable/disable gapfill.
It's based on my fork of slic3rpe, so with all my features & changes & the changes of slic3rPE.
I will merge it into the mainline slic3R if it's ok (and if i have the reward ;-p )

note: i didn't do a continuous mode for infill as it will takes me an other week to program and you didn't seems interested.

**Comment by gotthalmseder** [comment]
Wow, Merill! with the right wall-thickness the feature, as programmed right now, does a significant reduction of travel! Great! You get the reward ;) For this you can reach me in facebook chat (username: Manfred Gotthalmseder) or give me an other option to get in privat contact with you.

![image](https://user-images.githubusercontent.com/43730156/46628262-4b34df80-cb3d-11e8-96f3-74db0cbfbd43.png)

Right now I use Slic3r Prusa Edition 1.49.1 Win64  
But there is no big difference to the version you sent me. Your version even took my settings.

**Comment by gotthalmseder** [comment]
I tried the feature on a very complex object. But only the a part of the hight of the object got sliced. It might not be the new feature that trigger this. I allready had such problems without that feature. I guess it is the older Slic3r-Version that could be the reason. 
Another problem are these diagonal lines that occur. In this case they just occure if "detect bridging perimeters" is on. 
![complex_object](https://user-images.githubusercontent.com/43730156/46633253-71627b80-cb4d-11e8-809a-6532d57d093d.jpg)
You can download my testfile by clicking on this link: 
http://www.picsandpixels.at/testing.stl

**Comment by supermerill** [comment]
> It's limited to 20 layers (because it's only to test if it's ok, not a real version).

**Comment by gotthalmseder** [comment]
Oh, I am very happy, looking forward to print complex objects like that without the artefacts I have shown at the start. Seems you are a real good programmer! The labyrints the programm does look rather complex and it does it fast. 
The feature right now seems to work so well, that there should be a solution for all my functional parts as long as I draw things in a way that can be done well. 

The whole problem of slicing reminds me to a problem I know from my work in Photoshop. Programm-features to cut out objects are also confronted with the problem that there are many possible solutions. For there is no way that a programm finds the ideal solution itself, the features include the help of the user. So you click on areas that are not ok and the programm brings another solution. Maybe this would be an idea for the future of slicing as well. Because there will allways be some areas in an object you wish the programm would have done it different or would have done the seam on another place. But this is an idea for a new reward :-) The "continuous mode" is allready done! 

It would be nice to have continuous mode included in the latest version of slic3r. As I told you I use Slic3r Prusa Edition 1.49.1 Win64. Maybe you have good reasons to use an older verison, then just tell me.

**Comment by supermerill** [comment]
> As I told you I use Slic3r Prusa Edition 1.49.1 Win64

latest version of slic3rPE is 1.41.1

**Comment by gotthalmseder** [comment]
You are right, it was a typing error that I copied further on. My Edition is 1.40.1, so yours is newer. 
It seems all my questions are answered. To give me your billing-data it seems better to get in personal contact. You find my mailadress on www.picsandpixels.at or you reach me in facebook if you search for the username: Manfred Gotthalmseder.

**Comment by supermerill** [comment]
an other test build: https://mega.nz/#!oxQzHYZT!sS-I4b6SWSQ-upi0wURBrOjiIS0Xsqyjp7WGbElLV5o

This should be the final algorithm. I can't spot any bad crossing of perimeters, and it doesn't extrude when going over infill.

I have to test it irl, but i can't do that for the next weeks (because of a business trip). I will create a pr after test.
